design/kiota-e2e.md

```diff
@@ -0,0 +1,215 @@
+# End-to-end usage of kiota-generated SDKs in Javascript
+
+While working with the Kiota + Javascript teams, it became clear that we needed more context and provide a better vision on how we should be using a kiota-generated SDK. This design document aims at providing clarity and a better understanding on how developers will be using our SDKs and the underlying kiota packages.
+
+## Constraints
+
+Before we jump into the end-to-end walk-through, it's important to set some constraints.
+
+| Type      | Description                                                                                                                                   |
+| --------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
+| Platforms | Node : Current and Previous LTS (14 / 16)<br /> Web : Edge, Chrome, Firefox and Safari (Latest released version + immediate previous version) |
+| Modules   | Node : ESM (For preview) and potentially CJS (Based on feedback)<br /> Web : ESM                                                              |
+| Types     | Typings / Models should be available for both the core and the service libraries for Graph models                                             |
+
+## NodeJS e2e using the Service library
+
+### Concept
+
+This example is utilizing both the core and the service library. That means that as a developer, we want the full fluent API, allowing me discover APIs via the intellisence. Because the service library leverages the core library, we also want to highlight that we can leverage the core components directly (like the `.api()` method).
+
+### Install the Microsoft Graph Javascript Service Library
+
+There will be two versions of the service library for the Graph Javascript SDK. The default one will bring the representation of the v1.0 of Microsoft Graph. The second one will bring the representation of the beta of Microsoft Graph. This will be the only difference between these two libraries, as all the handcrafted code (delivering authentication, middlewares, tasks, etc.) will happen in the core library.
+
+The Microsoft Graph Javascript Service Library includes the following packages:
+
+-   `@microsoft/msgraph-sdk-javascript` ([npm](https://www.npmjs.com/package/@microsoft/msgraph-sdk-javascript-core)) - The service library for making fluent calls to Microsoft Graph v1.0.
+-   `@microsoft/msgraph-sdk-javascript-beta` ([npm](https://www.npmjs.com/package/@microsoft/msgraph-sdk-javascript-type)) - The service library for making fluent calls to Microsoft Graph beta.
+
+You can use [npm](https://www.npmjs.com) to install the Microsoft Graph Javascript Service Library :
+
+```Shell
+npm install @microsoft/msgraph-sdk-javascript --save
+npm install @microsoft/msgraph-sdk-javascript-beta --save-dev
+npm install @microsoft/kiota-authentication-azure --save
```

Comment:
What's the purpose of this special auth package? Is it just an example or it something that we'll expect developers to install and understand?

Reply:
this would be the authentication provider to use when you want to use Azure Identity. So most likely for SPAs, or nodeJS apps.
For SPFx/TeamsFx and other "platforms" this package would be different.

Reply:
Since MSAL already solved auth for SPAs and Node.js apps, couldn't we reuse what's there already?

Reply:
which doesn't solve for SPFx/TeamsFx/Electron and for other authentication mechanisms or libraries. The role of azure identity/msal is to unify the authentication experience for any API that lives behind Microsoft Identity Platform in scenarios where you don't already have a token. The role of those kiota wrappers is to abstract away authentication so the client is not tied to any specific implementation/scenario/mechanism. Two different concerns that go hand in hand.

Reply:
> The role of those kiota wrappers is to abstract away authentication so the client is not tied to any specific implementation/scenario/mechanism.

And that's exactly the problem because we're introducing yet another abstraction layer. So on top of identity APIs, we have MSAL, @azure/identity and now Kiota providers which are in total 4 different ways to do auth that developers need to know about and understand the benefits and trade-offs of each one of them.

I wonder if the concern about using different providers with SPFx, Teamsfx and other toolchains is relevant here, because these toolchains expose an authenticated version of the Graph client. Yes, it's important that the Graph SDK supports being instrumented with information coming from these toolchains, but developers using the Graph SDK will never see that. They will only see the special way of authenticating the SDK using the Kiota provider that applies only to the Graph SDK. If they use some other Azure SDK or raw REST calls, they will need to use MSAL/@azure identity there, which is a poor experience. What if support for authenticating toolchains was something that we handle internally but externally, for developers using the Graph SDK, they'd be able to use MSAL/@azure identity just as they do in any other Azure SDK or raw REST calls?

Reply:
> I didn't know you could mention an org on GitHub, I wonder whether it notifies anybody when you do that...

That wasn't my intention. It was meant to denote the `@azure/identity` package 😊